internal/bgp/bgp.go

```diff
@@ -525,8 +527,21 @@ type tcpmd5sig struct {
 // proper TCP MD5 options when the password is not empty. Works by manupulating
 // the low level FD's, skipping the net.Conn API as it has not hooks to set
 // the neccessary sockopts for TCP MD5.
-func dialMD5(ctx context.Context, addr, password string) (net.Conn, error) {
-	laddr, err := net.ResolveTCPAddr("tcp", "[::]:0")
+func dialMD5(ctx context.Context, addr string, srcAddr string, password string) (net.Conn, error) {
+	// If srcAddr exists on any of the local network interfaces, use it as the
+	// source address of the TCP socket. Otherwise, use the IPv6 unspecified
+	// address ("::") to let the kernel figure out the source address.
+	// NOTE: On Linux, "::" also includes "0.0.0.0" (all IPv4 addresses).
+	//
+	// When the source-address configuration option isn't set by the user,
+	// srcAddr becomes "<nil>", which makes localAddressExists() return false.
+	a := "[::]"
+	ifs, err := net.Interfaces()
+	if err == nil && localAddressExists(ifs, net.ParseIP(srcAddr)) {
```

Comment:
I'm not sure if the user sets this and the address doesn't exist, if we should fail instead. If we don't do that, then something that can easily happen is:
* I configure a src address A for a peer that is used only on one node
* Over time that peer applies to several nodes but all continues to work (src address A is ignored in all nodes, as not present)
* One of the nodes receives that IP address but doesn't connect to the peer (routes, whatever)
* By doing a very simple change (adding an ip to a node), you can break metallb on that node because of some legacy config that was left and metallb ignores without a warning nor anything.

Specifying a srcAddr is quite a clear indication to use it, I'd expect to use it or fail with an error, not to silently drop it.

Is there some reason to just not fail if the specified srcAddress can't be used? What do you think about raising an error instead?

Reply:
>Over time that peer applies to several nodes

We clearly say in the docs that source IP should only be used with per-node peers, i.e. peers which apply to one node only.

>Is there some reason to just not fail if the specified srcAddress can't be used? What do you think about raising an error instead?

We could do that. I've opted for a less "intrusive" behavior because I assumed we may still be able to establish a BGP session using the kernel-selected source IP.

Reply:
I think there might even be valid use cases for using the same source ip address from multiple hosts - think about link-local ipv6 addresses.

Reply:
> I think there might even be valid use cases for using the same source ip address from multiple hosts - think about link-local ipv6 addresses.

I agree, that's why I wrote "in most cases" [here](https://github.com/metallb/metallb/pull/902/files#diff-b3daf7076dbbe1b7fe39b03fe884a6a6fceef9859d489531069948319a6b38aaR266).

Reply:
I've changed the behavior so that we log an error if the user provides a malformed IP or an address which doesn't exist locally.